src/sha1.rs

```diff
@@ -22,118 +22,368 @@
  * the `reset` method.
  */
 
-
-use cryptoutil::{write_u32_be, read_u32v_be, add_bytes_to_bits, FixedBuffer, FixedBuffer64,
-    StandardPadding};
+use std::num::Int;
+use std::simd::u32x4;
 use digest::Digest;
+use cryptoutil::{write_u32_be, add_bytes_to_bits, FixedBuffer, FixedBuffer64, StandardPadding};
 
-/*
- * A SHA-1 implementation derived from Paul E. Jones's reference
- * implementation, which is written for clarity, not speed. At some
- * point this will want to be rewritten.
- */
+const STATE_LEN: usize = 5;
+const BLOCK_LEN: usize = 16;
 
-// Some unexported constants
-const DIGEST_BUF_LEN: usize = 5;
-const WORK_BUF_LEN: usize = 80;
 const K0: u32 = 0x5A827999u32;
 const K1: u32 = 0x6ED9EBA1u32;
 const K2: u32 = 0x8F1BBCDCu32;
 const K3: u32 = 0xCA62C1D6u32;
 
-/// Structure representing the state of a Sha1 computation
-#[derive(Copy)]
-pub struct Sha1 {
-    h: [u32; DIGEST_BUF_LEN],
-    length_bits: u64,
-    buffer: FixedBuffer64,
-    computed: bool,
+/*
+ *  /// Emulates `llvm.arm.neon.sha1h` intrinsic.
+ *  ///
+ *  /// (The letter 'H' might stand for half, maybe?)
+ *  #[inline]
+ *  fn sha1h(a: u32) -> u32 {
+ *      a.rotate_left(30)
+ *  }
+ *
+ *  /// Emulates `llvm.arm.neon.sha1su0` intrinsic.
+ *  #[inline]
+ *  fn sha1su0(a: u32x4, b: u32x4, c: u32x4) -> u32x4 {
+ *      sha1msg1(a, b) ^ c
+ *  }
+ *
+ *  /// Emulates `llvm.arm.neon.sha1su1` intrinsic.
+ *  #[inline]
+ *  fn sha1su1(a: u32x4, b: u32x4) -> u32x4 {
+ *      sha1msg2(a, b)
+ *  }
+ *
+ *  /// Emulates `llvm.arm.neon.sha1c` intrinsic.
+ *  #[inline]
+ *  fn sha1c(abcd: u32x4, e: u32, msg: u32x4) -> u32x4 {
+ *      sha1rnds4c(abcd, sha1stadd(e, msg))
+ *  }
+ *
+ *  /// Emulates `llvm.arm.neon.sha1p` intrinsic.
+ *  #[inline]
+ *  fn sha1p(abcd: u32x4, e: u32, msg: u32x4) -> u32x4 {
+ *      sha1rnds4p(abcd, sha1stadd(e, msg))
+ *  }
+ *
+ *  /// Emulates `llvm.arm.neon.sha1m` intrinsic.
+ *  #[inline]
+ *  fn sha1m(abcd: u32x4, e: u32, msg: u32x4) -> u32x4 {
+ *      sha1rnds4m(abcd, sha1stadd(e, msg))
+ *  }
+ */
```

Comment:
Nope, I'll remove it.

I think it is easier to implement x86 intructions with arm intructions, than it is to implement arm instructions with x86 instructions, so I believe that the x86 intrinsic functions are a better interface to present to the public.